Add IFrame for loading apps

sohkai opened the pull request:
![cool iframe](https://thumbs.gfycat.com/BriefElectricIbisbill-max-14mb.gif)

[See higher res](https://gfycat.com/BriefElectricIbisbill)

Fixes #60 (see 5b3a422), fixes #69 (see 8adee4d).

Currently every app is set to load a demo app (see 4454eb3) that's built along with the app. In the future, we can use a similar workflow to migrate the other native apps (e.g. Settings, Permissions) to a similar structure so that they're built alongside the wrapper as a separate project. Once we've built the other apps, we can also wholesale remove the demo app.

izqui commented:
This is great!

izqui commented on src/components/App/AppIFrame.js:
```diff
@@ -0,0 +1,140 @@
+// `sandbox` works like a whitelist: by default, almost every functionality is restricted.
+const SANDBOX = [
+  // Allows new browsing contexts (e.g. new windows or tabs) generated by the
+  // iframe to escape the sandbox restrictions. Otherwise, new contexts would
```
👍 can we do some tests to make sure that's the case and it is impossible for it to run any JS in the parent

sohkai replied:
I tested it spuriously while playing around with iframes earlier, but thinking about it, we could put up some random tests under a repo in aragonlabs.

It should be noted that the demo app isn't able to load any assets (e.g. fonts; you'll see a bunch of errors in the console) from the same origin because of the origin sandbox that's been set 😅.

izqui replied:
In the future this should be part this project's test suite, but I imagine we would need to run it against a headless browser or something. Nothing too urgent, but this could be an attack vector.

izqui replied:
What @onbjerg says in the comment below could be a cool test case, make a super malicious Aragon app that tries to break the sandbox and do nasty stuff ;)

sohkai replied:
Added https://github.com/aragon/aragon/issues/87.

izqui commented on src/components/App/AppIFrame.js:
```diff
@@ -0,0 +1,140 @@
+  //  - 'allow-top-navigation':
+  //       Leaving this disabled disallows the iframe from navigating the
+  //       Dapp's context (e.g. to a malicious page)
+].join(' ')
```
I want to somehow audit these permissions to make sure we are good :)

onbjerg replied:
I did this when choosing them for my old example wrapper by trying to break out of the iframe, these are fine.

We can always make a malicious app we can test against :)

izqui commented:
Regarding loading content from IPFS, are we just using public gateways for this? Load `ipfs.io/ipfs/hash` or however it is?

onbjerg commented:
@izqui yeah but we should make the gateway configurable such that we could also use localhost if we wanted to

izqui commented:
👍

El lun., 8 ene. 2018 a las 11:52, Oliver (<notifications@github.com>)
escribió:

> @izqui <https://github.com/izqui> yeah but we should make the gateway
> configurable such that we could also use localhost if we wanted to
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/aragon/aragon/pull/83#issuecomment-355935219>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAbTYCn5vI2CnutlhrQ7IKYr-kZCSOEPks5tIfOBgaJpZM4RWEgv>
> .
>

sohkai commented on src/components/App/AppLoadingProgressBar.js:
```diff
@@ -0,0 +1,25 @@
+import React from 'react'
+import { Line } from 'rc-progress'
```
I choose `rc-progress` mostly because it was a super light svg implementation, but thinking about it, we could likely just use a `div` with background colour and width set 👍.

sohkai commented:
> For the bar progress, maybe we could try to use our fast (or even faster) spring animation to make it smoother?

👍 I'll try playing around with this.

> The bar sometimes disappear before completion: maybe we could try to animate it before it disappears 

Hmm, the bar definitely shouldn't be disappearing before completing, unless the dapp gets navigated back to the Home app (it should be waiting ~1s after loading before being reset). Are you seeing that?

I'll try a quick fade-in and fade-out animation with the spring though.

> On Firefox only, there is some flickering when an app is loaded

Ah, I can confirm this is due to the fonts being blocked from the iframe sandbox. For some reason, Firefox flickers when it can't load the font. If you add `allow-same-origin` to the iframe sandbox, it shouldn't happen anymore.

I did find a weird race condition with the progress bar moving backwards when it starts though :(.

sohkai commented:
@bpierre [See new loading gif](https://gfycat.com/backthickcommongonolek) (delayed the `onLoad` by 10s for more fudging). I've removed `rc-progress` for our own implementation with `div`s.

I've added a fade-in when the progress bar appears, but I can't come up with a good way to add a fade-out animation in a safe (non-present race condition) way without a larger re-architecture. Let me know if I'm missing something really obvious with `react-motion` though...

Again, let me know if you see the progress bar disappearing before it completes when not re-navigating.

bpierre commented:
> Hmm, the bar definitely shouldn't be disappearing before completing, unless the dapp gets navigated back to the Home app (it should be waiting ~1s after loading before being reset). Are you seeing that?

Yes, not on the first load, but when I alternate between e.g. “Home” and “Tokens”, it sometimes happen.

About the 1s pause (sorry, I only realize now it was intentional):

I agree that we need to show the progress state as completed, but I think 1s is a bit too long. The app is fully loaded at this point, but the presence of the loading bar make it feel like the app is not ready.

A solution could be to start the hiding transition right at this point, instead of pausing. I think this transition could be super fast, because the disappearance of the bar is enough to understand that the app is ready. What do you think?

> Ah, I can confirm this is due to the fonts being blocked from the iframe sandbox. For some reason, Firefox flickers when it can't load the font. If you add allow-same-origin to the iframe sandbox, it shouldn't happen anymore.

Indeed! I forked react-scripts to have the fonts in development (using `npm link` for now, let’s see if the PR is accepted) https://github.com/facebookincubator/create-react-app/pull/3759

bpierre commented:
> Again, let me know if you see the progress bar disappearing before it completes when not re-navigating.

Fixed with the last commit! 👍

bpierre commented on src/components/App/AppIFrame.js:
```diff
@@ -0,0 +1,154 @@
+          }}
+          sandbox={SANDBOX}
+          style={{
+            display: show ? 'initial' : 'none',
```
`initial` should be `block` here, otherwise it will revert back to `inline` (the default display for `iframe`) and a vertical scrollbar will appear.

sohkai commented:
Pausing for 250ms: https://gfycat.com/impeccabledopeyiriomotecat
No pause: https://gfycat.com/CheapSillyFlea

I'm leaning on the no pause as well (c0a99a0) (but those gyfcat names :rofl:).

> I forked react-scripts to have the fonts in development

Awesome 🏆!
